pkg/disttask/framework/scheduler/balancer.go

```diff
@@ -0,0 +1,218 @@
+	executorPendingCnts := make(map[string]int, len(adjustedNodes))
+	for _, node := range adjustedNodes {
+		executorSubtasks[node] = make([]*proto.Subtask, 0, averageSubtaskCnt+1)
+	}
+	for _, subtask := range subtasks {
+		// put running subtask in the front of slice.
+		// if subtask fail-over, it's possible that there are multiple running
+		// subtasks for one task executor.
+		if subtask.State == proto.SubtaskStateRunning {
+			executorSubtasks[subtask.ExecID] = append([]*proto.Subtask{subtask}, executorSubtasks[subtask.ExecID]...)
+		} else {
+			executorSubtasks[subtask.ExecID] = append(executorSubtasks[subtask.ExecID], subtask)
+			executorPendingCnts[subtask.ExecID]++
+		}
+	}
+
+	subtasksNeedSchedule := make([]*proto.Subtask, 0)
+	remainder := averageSubtaskRemainder
+	executorWithOneMoreSubtask := make(map[string]struct{}, remainder)
+	for k, v := range executorSubtasks {
+		if _, ok := adjustedNodeMap[k]; !ok {
+			// dead node or not have enough slots
+			subtasksNeedSchedule = append(subtasksNeedSchedule, v...)
+			delete(executorSubtasks, k)
+			continue
+		}
+		if remainder > 0 {
+			// first remainder nodes will get 1 more subtask.
+			if len(v) >= averageSubtaskCnt+1 {
+				needScheduleCnt := len(v) - (averageSubtaskCnt + 1)
+				// running subtasks are never balanced.
+				needScheduleCnt = min(executorPendingCnts[k], needScheduleCnt)
+				subtasksNeedSchedule = append(subtasksNeedSchedule, v[len(v)-needScheduleCnt:]...)
+				executorSubtasks[k] = v[:len(v)-needScheduleCnt]
+
+				executorWithOneMoreSubtask[k] = struct{}{}
+				remainder--
+			}
+		} else if len(v) > averageSubtaskCnt {
+			// running subtasks are never balanced.
+			cnt := min(executorPendingCnts[k], len(v)-averageSubtaskCnt)
+			subtasksNeedSchedule = append(subtasksNeedSchedule, v[len(v)-cnt:]...)
+			executorSubtasks[k] = v[:len(v)-cnt]
+		}
+	}
+	if len(subtasksNeedSchedule) == 0 {
+		return nil
+	}
+
+	for i := 0; i < len(adjustedNodes) && remainder > 0; i++ {
+		if _, ok := executorWithOneMoreSubtask[adjustedNodes[i]]; !ok {
+			executorWithOneMoreSubtask[adjustedNodes[i]] = struct{}{}
+			remainder--
+		}
+	}
+
+	fillIdx := 0
+	for _, node := range adjustedNodes {
+		sts := executorSubtasks[node]
+		targetSubtaskCnt := averageSubtaskCnt
+		if _, ok := executorWithOneMoreSubtask[node]; ok {
+			targetSubtaskCnt = averageSubtaskCnt + 1
+		}
+		for i := len(sts); i < targetSubtaskCnt && fillIdx < len(subtasksNeedSchedule); i++ {
+			subtasksNeedSchedule[fillIdx].ExecID = node
+			fillIdx++
+		}
+	}
+
+	if err = b.taskMgr.UpdateSubtasksExecIDs(ctx, subtasksNeedSchedule); err != nil {
+		return err
+	}
+	logutil.BgLogger().Info("balance subtasks", zap.Stringers("subtasks", subtasksNeedSchedule))
+	return nil
+}
+
+func (b *balancer) updateUsedNodes(subtasks []*proto.Subtask) {
+	used := make(map[string]int, len(b.currUsedSlots))
+	for _, st := range subtasks {
+		if _, ok := used[st.ExecID]; !ok {
+			used[st.ExecID] = st.Concurrency
+		}
+	}
```

Comment:
Can you add some comment to explain why `currUsedSlots` only calculate one subtask for each node?